### PR TITLE
fix utf16 offset handling

### PR DESCRIPTION
**Pull request by ZacLN**
I think this should do it, supersedes #419. I think before we were essentially treating the character index as UTF32.

Thanks to @non-Jedi for finding this issue along with all the background ...

### CODE REVIEW

File: src/document.jl [hunk 1]
```diff
@@ -60,6 +60,9 @@ function get_offset(doc::Document, line::Integer, character::Integer)
     offset = line_offsets[line + 1]
     while character > 0
         offset = nextind(doc._content, offset)
+        if nextind(get_text(doc), offset) - offset > 2
```
Comment (davidanthoff) [inline src/document.jl]: So this seems the wrong cut-off? UTF-8 uses more than 2 codepoints for anything above U+0800, but for UTF-16 the cutoff from 1 to 2 codepoints is 0x010000, I believe. So I think we need to capture the case where a character is above 0x010000, right? Also, wouldn't it be easier to directly check for the value of the char, like I did at https://github.com/julia-vscode/LanguageServer.jl/pull/419/files#diff-434665f0bad6ba3f5bf14654c456fa35R61?

@StefanKarpinski, can you help? I feel on very shaky ground whenever it comes to anything UTF :)

File: src/document.jl [hunk 2]
```diff
@@ -119,10 +122,14 @@ function get_position_at(doc::Document, offset::Integer)
     offset > sizeof(doc._content) && error("offset[$offset] > sizeof(content)[$(sizeof(doc._content))]")
     line_offsets = get_line_offsets(doc)
     line, ind = get_line_of(doc._line_offsets, offset)
+    ind = nextind(get_text(doc), ind)
     char = 0
-    while offset > ind
-        ind = nextind(doc._content, ind)
+    while offset >= ind
+        if nextind(get_text(doc), ind) - ind > 2
```
Comment (davidanthoff) [inline src/document.jl]: Same story here, I don't think having more than 2 codepoints in UTF-8 is the right cut-off, if I understand UTF-8 and UTF-16 correctly.

File: src/document.jl (hunk 1 above)
Comment (StefanKarpinski) [reply src/document.jl]: Yes, I think 0x010000 is right.

Comment (ZacLN) [reply src/document.jl]: Gotcha, I've altered it to use an IOBuffer and explicitly check at those cutoffs

**Comment by ZacLN** [comment]
For info - `mark_errors` uses exactly the same approach as `get_position_at` but handles (sorted) batches of offsets

**Comment by StefanKarpinski** [comment]
Can you explain a bit more clearly what the expectations on the inputs and outputs are in terms of encoding and what's computed?

File: src/document.jl [hunk 3]
```diff
@@ -56,13 +56,26 @@ This takes 0 based line/char inputs. Corresponding functions are available for
...
+        end
+    end
+    if UInt32(c) < 0x0080
+        return position(io)
+    elseif UInt32(c) < 0x0800
+        return position(io) - 1
+    elseif UInt32(c) < 0x010000
+        return position(io) - 2
+    else
+        return position(io) - 3
```
[the remark this answers is not in the document]
Comment (ZacLN) [reply src/document.jl]: The position of the IO is that of the last byte of the utf8 representation of the char, hence this offset.

File: src/document.jl [hunk 4]
```diff
@@ -119,12 +132,18 @@ function get_position_at(doc::Document, offset::Integer)
...
+    while offset > position(io)
+        c = read(io, Char)
+        character += 1
+        if UInt32(c) >= 0x010000
+            character += 1
+        end
     end
-    return line - 1, char
+    close(io)
+    return line - 1, character
```
[the remark this answers is not in the document]
Comment (ZacLN) [reply src/document.jl]: Your PR wasn't giving the answers I wanted when I tested it! (Agree its simpler)

File: src/document.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidanthoff) [reply src/document.jl]: Ah, got it!

File: src/document.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidanthoff) [reply src/document.jl]: Hehe, yes, tests failed :) Lets go with this for now, and then we can still try to make this simpler later.

**Comment by ZacLN** [comment]
Super, I'd rather keep it obvious (to me) how it works incase further issues arise immediately

**Comment by davidanthoff** [comment]
> I'd rather keep it obvious (to me) how it works incase further issues arise immediately

Yep, that seems most important :)

So I'd say you merge this here, and then we try to fix the tests on `master`? And then I tag new versions of all packages and release a first release candidate?

**Comment by ZacLN** [comment]
Yep sounds like a plan

**Pull request merged**